identity-actor/src/actor/system.rs

```diff
@@ -0,0 +1,235 @@
+  state: Arc<SystemState>,
+}
+
+impl Clone for System {
+  fn clone(&self) -> Self {
+    Self {
+      commander: self.commander.clone(),
+      state: self.state.clone(),
+    }
+  }
+}
+
+impl System {
+  pub(crate) fn new(commander: NetCommander, state: Arc<SystemState>) -> System {
+    Self { commander, state }
+  }
+
+  pub(crate) fn state(&self) -> &SystemState {
+    self.state.as_ref()
+  }
+
+  /// Returns the [`PeerId`] that other peers can securely identify this actor with.
+  pub fn peer_id(&self) -> PeerId {
+    self.state().peer_id
+  }
+
+  pub fn commander_mut(&mut self) -> &mut NetCommander {
+    &mut self.commander
+  }
+
+  pub fn commander(&self) -> &NetCommander {
+    &self.commander
+  }
+
+  /// Start listening on the given `address`. Returns the first address that the actor started listening on, which may
+  /// be different from `address` itself, e.g. when passing addresses like `/ip4/0.0.0.0/tcp/0`. Even when passing a
+  /// single address, multiple addresses may end up being listened on. To obtain all those addresses, use
+  /// [`System::addresses`]. Note that even when the same address is passed, the returned address is not deterministic,
+  /// and should thus not be relied upon.
+  pub async fn start_listening(&mut self, address: Multiaddr) -> ActorResult<Multiaddr> {
+    self.commander_mut().start_listening(address).await
+  }
+
+  /// Return all addresses that are currently being listened on.
+  pub async fn addresses(&mut self) -> ActorResult<Vec<Multiaddr>> {
+    self.commander_mut().get_addresses().await
+  }
+
+  /// Shut this actor down. This will break the event loop in the background immediately,
+  /// returning an error for all current handlers that interact with their copy of the
+  /// actor or those waiting on messages. The actor will thus stop listening on all addresses.
+  ///
+  /// Calling this and other methods, which interact with the event loop, on an actor that was shutdown
+  /// will return [`Error::Shutdown`].
+  pub async fn shutdown(mut self) -> ActorResult<()> {
+    // Consuming self drops the internal commander. If this is the last copy of the commander,
+    // the event loop will break as a result. However, if copies exist, such as in running handlers,
+    // this function will return while the event loop keeps running. Ideally we could then join on the background task
+    // to wait for all handlers to finish gracefully. However, not all spawn functions return a JoinHandle,
+    // such as wasm_bindgen_futures::spawn_local. The current alternative is to use a non-graceful exit,
+    // which breaks the event loop immediately and returns an error through all open channels that require a result.
+    self.commander_mut().shutdown().await
+  }
+
+  /// Associate the given `peer_id` with an `address`. This needs to be done before sending a
+  /// request to this [`PeerId`].
+  pub async fn add_address(&mut self, peer_id: PeerId, address: Multiaddr) -> ActorResult<()> {
```

Comment:
Consider explaining how this is different from `start_listening` in the documentation.

Reply:
Thanks. Maybe `add_peer_address` is even more appropriate, given that it associates a `peer_id` with an `address`? Otherwise, I'm happy with your suggestion.

Reply:
`add_peer_address` is also nice. Use whichever you think is best.